core/native_test.go

```diff
@@ -0,0 +1,348 @@
+	txL := make([]*types.CrossShardTransactionDeposit, 0)
+	txL = append(txL, &deposit)
+	txList := types.CrossShardTransactionDepositList{TXList: txL}
+	shardState0.AddCrossShardTxListByMinorBlockHash(b1.Header().Hash(), txList)
+	//Create a root block containing the block with the x-shard tx
+	rootBlock = shardState0.GetRootTip().CreateBlockToAppend(nil, nil, nil, nil, nil)
+	rootBlock.AddMinorBlockHeader(shardState0.CurrentBlock().Header())
+	rootBlock.AddMinorBlockHeader(shardState1.CurrentBlock().Header())
+	rootBlock.Finalize(nil, nil, common.EmptyHash)
+	shardState0.AddRootBlock(rootBlock)
+	//Add b0 and make sure all x-shard tx's are added
+	b2, _ := shardState0.CreateBlockToMine(nil, &acc3, nil, nil, nil)
+	shardState0.FinalizeAndAddBlock(b2)
+	balance1, _ := shardState0.GetBalance(acc1.Recipient, nil)
+	balance3, _ := shardState0.GetBalance(acc1.Recipient, nil)
+	assert.Equal(t, balance1.GetTokenBalance(QETHXX), new(big.Int).Add(big.NewInt(999999), big.NewInt(888888)))
+	reward := new(big.Int).Add(testShardCoinbaseAmount, big.NewInt(18000)).Uint64()
+	assert.Equal(t, balance3.GetTokenBalance(QKC), afterTax(reward, shardState0))
+	evmState, _ := shardState0.State()
+	//X-shard gas used
+	assert.Equal(t, evmState.GetGasUsed(), big.NewInt(9000))
+}
+
+func TestXshardNativeTokenGasSent(t *testing.T) {
+	QETHXX := common.TokenIDEncode("QETHXX")
+	QKC := common.TokenIDEncode("QKC")
+	id1, _ := account.CreatRandomIdentity()
+	acc1 := account.CreatAddressFromIdentity(id1, 0)
+	acc2 := account.CreatAddressFromIdentity(id1, 1)
+	acc3 := account.CreatEmptyAddress(0)
+	testGenesisMinorTokenBalance["QETHXX"] = big.NewInt(999999)
+	testGenesisMinorTokenBalance["QKC"] = big.NewInt(10000000)
+	env := getTestEnv(&acc1, nil, nil, nil, nil, nil)
+	shardId := uint32(0)
+	shardState := createDefaultShardState(env, &shardId, nil, nil, nil)
+	env1 := getTestEnv(&acc1, nil, nil, nil, nil, nil)
+	shardId1 := uint32(1)
+	testGenesisMinorTokenBalance = make(map[string]*big.Int)
+	shardState1 := createDefaultShardState(env1, &shardId1, nil, nil, nil)
+	rootBlock := shardState.GetRootTip().CreateBlockToAppend(nil, nil, nil, nil, nil)
+	rootBlock.AddMinorBlockHeader(shardState.CurrentBlock().Header())
+	rootBlock.AddMinorBlockHeader(shardState1.CurrentBlock().Header())
+	rootBlock.Finalize(nil, nil, common.EmptyHash)
+	shardState.AddRootBlock(rootBlock)
+	val := big.NewInt(888888)
+	gas := new(big.Int).Add(big.NewInt(9000), big.NewInt(21000)).Uint64()
+	tx := createTransferTransaction(shardState, id1.GetKey().Bytes(), acc1, acc2, val, &gas, nil, nil, nil, &QETHXX, &QETHXX)
+	shardState.AddTx(tx)
+	b1, _ := shardState.CreateBlockToMine(nil, &acc3, nil, nil, nil)
+	assert.Equal(t, len(b1.Transactions()), 1)
+	evmState, _ := shardState.State()
+	assert.Equal(t, evmState.GetGasUsed(), 0)
+	shardState.FinalizeAndAddBlock(b1)
+	assert.Equal(t, len(evmState.GetXShardList()), 1)
```

Comment:
not equal

Reply:
https://github.com/QuarkChain/pyquarkchain/blob/08aa932343780279254cdd30dd665a090f931d46/quarkchain/cluster/tests/test_native_token.py#L447